pkg/action/diff.go

```diff
@@ -164,55 +181,93 @@ func fileDestination(ctx context.Context, c bincapz.Config, fr, tr *bincapz.File
 	}
 }
 
-type diffReports struct {
-	Added     string
-	AddedFR   *bincapz.FileReport
-	Removed   string
-	RemovedFR *bincapz.FileReport
+// filterMap filters orderedmap pairs by checking for matches against a slice of compiled regular expression patterns.
+func filterMap(om *orderedmap.OrderedMap[string, *bincapz.FileReport], ps []*regexp.Regexp, c chan<- *orderedmap.Pair[string, *bincapz.FileReport], wg *sync.WaitGroup) {
+	defer wg.Done()
+	for pair := om.Oldest(); pair != nil; pair = pair.Next() {
+		for _, pattern := range ps {
+			if match := pattern.FindString(filepath.Base(pair.Key)); match != "" {
+				c <- pair
+			}
+		}
+	}
 }
 
-// combineReports builds a flattened slice of added and removed paths and their respective file reports.
-func combineReports(d *bincapz.DiffReport) []diffReports {
-	diffs := make(chan diffReports)
+// combine iterates over the removed and added channels to create a diff report to store in the combined channel.
+func combine(removed, added <-chan *orderedmap.Pair[string, *bincapz.FileReport], combined chan<- bincapz.CombinedReport) {
+	for r := range removed {
+		for a := range added {
+			score := levenshtein.Match(r.Key, a.Key, levenshtein.NewParams())
+			if score < 0.9 {
+				continue
+			}
+			combined <- bincapz.CombinedReport{
+				Added:     a.Key,
+				AddedFR:   a.Value,
+				Removed:   r.Key,
+				RemovedFR: r.Value,
+				Score:     score,
+			}
+		}
+	}
+}
+
+// combineReports orchestrates the population of the diffs channel with relevant diffReports.
+func combineReports(c bincapz.Config, d *bincapz.DiffReport, combined chan<- bincapz.CombinedReport) {
+	defer close(combined)
+
+	maxConcurrency := c.Concurrency
+	if maxConcurrency < 1 {
+		maxConcurrency = 1
+	}
+
 	var wg sync.WaitGroup
 
-	for removed := d.Removed.Oldest(); removed != nil; removed = removed.Next() {
-		wg.Add(1)
-		go func(path string, fr *bincapz.FileReport) {
-			defer wg.Done()
-			for added := d.Added.Oldest(); added != nil; added = added.Next() {
-				diffs <- diffReports{
-					Added:     added.Key,
-					AddedFR:   added.Value,
-					Removed:   path,
-					RemovedFR: fr,
-				}
-			}
-		}(removed.Key, removed.Value)
+	// Patterns we care about when handling diffs
+	patterns := []string{
+		`^[\w.-]+\.so$`,
+		`^.+-.*-r\d+\.spdx\.json$`,
+	}
+
+	ps := make([]*regexp.Regexp, len(patterns))
+	for i, pattern := range patterns {
+		ps[i] = regexp.MustCompile(pattern)
 	}
 
+	// Build two channels with filtered paths to iterate through in the worker pool
+	removed := make(chan *orderedmap.Pair[string, *bincapz.FileReport], d.Removed.Len())
+	added := make(chan *orderedmap.Pair[string, *bincapz.FileReport], d.Added.Len())
+
+	wg.Add(1)
```

Comment:
Similar to scanning each of the paths concurrently, we'll also filter the resulting removed/added maps concurrently.

Reply:
Prior to me opening this PR, I had added a worker pool pattern to handle the maps in chunks to reduce memory usage, but since we're filtering the maps there's not a need for that any longer.